mnd/mnd/integration/mims/mims_service.py

```diff
@@ -0,0 +1,119 @@
+from collections import namedtuple
+import logging
+import requests
+import uuid
+
+from django.shortcuts import reverse
+
+from .mims_api import MIMSApi
+from .mims_cache import (
+    CMIInfo, ProductInfo, get_product, get_cmi_by_product,
+    get_cmi_info, search_cache, update_cache,
+    update_cache_entry, update_cmi_cache
+)
+
+ProductSearchResult = namedtuple('ProductSearchResult', 'id value activeIngredient')
+
+logger = logging.getLogger(__name__)
+
+api = MIMSApi()
+
+
+def _is_valid_uuid(input_str):
+    try:
+        uuid.UUID(input_str)
+    except Exception:
+        return False
+    return True
+
+
+def mims_product_search(product):
+
+    def active_ingredient(result):
+        return ", ".join(result.split(" + "))
+
+    if product and product.strip() != '' and len(product) >= 4:
+        cached = search_cache(product)
+        if cached:
+            return [ProductSearchResult(r.id, r.name, r.activeIngredient) for r in cached]
+
+        result = api.search_product(product)
```

Comment:
I see there are pagination fields in the API. Do we need to consider the edge case where a common drug has so many variants that turn up in a search, that we need to paginate over it to provide a complete list to the user?

Or maybe there needs to be a visual indication that the results aren't complete if there are more pages?

Reply:
Added pagination support when searching products.